### ISSUE TITLE
acquireTokenSilent: No cached accounts found for the supplied homeAccountId and clientId

### ISSUE THREAD

**Issue by pabloromeu** [AzureAD/microsoft-authentication-library-for-android#1033]
**Describe the bug**
I am using a wrapper over this library (https://github.com/stashenergy/react-native-msal) to build a react native App. This app works fine on iOS, but after a successful `acquireToken`call, fails to `acquireTokenSilent`  on Android with error: 
```
com.microsoft.identity.client.exception.MsalClientException: No cached accounts found for the supplied homeAccountId and clientId
```
I have debugged the error and it is located at this MSAL library.

**Smartphone (please complete the following information):**
 - Device: Using Android Emulator Pixel 2
 - Android Version: API 28
 - Browser: Chrome
 - MSAL Version: initially 1.2.0, but also tried 1.3.0 and 1.4.1

**Stacktrace**
```
W/System.err: com.microsoft.identity.client.exception.MsalClientException: No cached accounts found for the supplied homeAccountId and clientId
W/System.err:     at com.microsoft.identity.client.internal.controllers.MsalExceptionAdapter.msalExceptionFromBaseException(MsalExceptionAdapter.java:51)
        at com.microsoft.identity.client.PublicClientApplication$9.onError(PublicClientApplication.java:1630)
W/System.err:     at com.microsoft.identity.client.PublicClientApplication$9.onError(PublicClientApplication.java:1621)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$2.run(CommandDispatcher.java:185)
        at android.os.Handler.handleCallback(Handler.java:873)
W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:99)
W/System.err:     at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
    Caused by: com.microsoft.identity.common.exception.ClientException: No cached accounts found for the supplied homeAccountId and clientId
W/System.err:     at com.microsoft.identity.common.internal.controllers.BaseController.getCachedAccountRecord(BaseController.java:455)
        at com.microsoft.identity.client.internal.controllers.LocalMSALController.acquireTokenSilent(LocalMSALController.java:251)
        at com.microsoft.identity.common.internal.controllers.TokenCommand.execute(TokenCommand.java:70)
        at com.microsoft.identity.common.internal.controllers.TokenCommand.execute(TokenCommand.java:39)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher.executeCommand(CommandDispatcher.java:146)
W/System.err:     at com.microsoft.identity.common.internal.controllers.CommandDispatcher.access$200(CommandDispatcher.java:60)
W/System.err:     at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:103)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
```
**To Reproduce**

I am using a the library to authenticate through Azure B2C.

My msal_config.json file looks like: 

```
{
  "client_id" : "792...,
  "redirect_uri" : "msauth://com..../QyXa...",
  "account_mode":"MULTIPLE",
  "broker_redirect_uri_registered": false,
  "authorities" : [
    {
      "type": "B2C",
      "authority_url": "https://....b2clogin.com/tfp/....onmicrosoft.com/B2C_1A_signup_signin_...",
      "default": true
    }
  ]
}
```

I do a first call to `acquireToken` and receive this -**_notice there is no idToken at the response_**-:

```
{
  "account": {
    "identifier": "GMJ2...-b2c_1a_signup_signin_whatever,
    "username": "Missing from the token response"
  },
  "tenantId": "a33d....,
  "authority": "https://whatever.b2clogin.com/tfp/whatever.onmicrosoft.com/B2C_1A_signup_signin_whatever",
  "scopes": [
    "https://whatever.onmicrosoft.com/79.../User.Read",
    "openid",
    "profile",
    "offline_access"
  ],
  "idToken": null, 
  "expiresOn": "1588277155000",
  "accessToken": " a full access token"
}```
```

after that the `acquireTokenSilent` call receives these parameters:

```
{
   authority: "https://....b2clogin.com/tfp/....onmicrosoft.com/B2C_1A_signup_signin_",
   scopes: ["aScope"],
   accountIdentifier: this.auth.account.identifier,
   forceRefresh: true
}
```

And it fails with: 
```
com.microsoft.identity.client.exception.MsalClientException: No cached accounts found for the supplied homeAccountId and clientId
```

**Expected behavior**
To renews the token

**Actual Behavior**
It fails to renew the token

**Screenshots**
N/A

**Additional context**

I found that the library is trying to do some kind of idToken catching -at `MsalOauth2TokenCache`-, and maybe the problem could be related to it. 

I also filled a bug on the react-native wrapper library with the information I have included here, but as I said, I think it has to do with the cache of this library: https://github.com/stashenergy/react-native-msal/issues/11

**Comment by iambmelt** [comment]
As with using any wrapper around our library YMMV - at a glance, I would check that your wrapper is passing the account object into `acquireTokenSilent`. From the error message, [this doesn't look to be the case](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java#L485-L490).

I suggest taking a look at our [sample app with B2C](https://github.com/Azure-Samples/ms-identity-android-java#b2cmodefragment-class) -- it's worth noting that account object from 1 policy cannot be passed haphazardly to another policy. Because each policy is considered to be its own authority,  only accounts belonging to that policy can be passed into `acquireTokenSilent()`.

**Comment by pabloromeu** [comment]
Thanks @iambmelt for the hint. I paste the response of the `acquireToken` call. It's strange that idToken is null. On iOS the idToken is returned:

![response](https://user-images.githubusercontent.com/3224197/81141899-42e99400-8f6e-11ea-84d3-c5242fded859.png)

Account is indeed passed, but obviously, idToken is not there. Parameters passed to `acquireTokenSilentAsync`:

![Untitled](https://user-images.githubusercontent.com/3224197/81142038-b9869180-8f6e-11ea-836c-e9e390f0dcf6.png)

**Comment by pabloromeu** [comment]
FYI, my users in B2C are federated. I don't know whether that helps, but...

**Comment by joshfriend** [comment]
> account object from 1 policy cannot be passed haphazardly to another policy

I'm having this error as well, presumably because I tried to use `configuration.defaultAuthority.authorityURL` to acquire a token for a user that is not part of my AAD tenant. Older versions of this library had a `Authority.getAuthorityFromAccount` method that could tell you the right authority to use and was called by default if no authority was provided in the `AcquireTokenSilentParams`.

Whats the reason in the new version for not guessing a sane default authority URL and forcing the user to figure it out for themselves? Right now I'm really confused how I get users from other tenants working again.

**Comment by iambmelt** [comment]
> FYI, my users in B2C are federated. I don't know whether that helps, but...

@pabloromeu Are you continuing to have this issue? If yes, we have provided a B2C sample here that you can clone/run to observe how to authenticate a user across policies:
https://github.com/Azure-Samples/ms-identity-android-java#b2cmodefragment-class

**Comment by josmithua** [comment]
@joshfriend Hi and sorry I didn't see your comment before now. Feel free to open an issue in the react-native-msal repo and we can discuss there.

**Comment by icc-romeu** [comment]
> > FYI, my users in B2C are federated. I don't know whether that helps, but...
> 
> @pabloromeu Are you continuing to have this issue? If yes, we have provided a B2C sample here that you can clone/run to observe how to authenticate a user across policies:
> https://github.com/Azure-Samples/ms-identity-android-java#b2cmodefragment-class

No It was fixed. We contacted Azure Support and they came back with a modification of our Azure B2C configuration. After the change it worked. I am sorry I cannot give you a hint on what we changed because I develop the mobile part, not the Azure one.

This can be closed. Thank you.

**Comment by iambmelt** [comment]
Thanks for the update @icc-romeu -- glad to hear you were unjammed

**Comment by mtapiero** [comment]
@icc-romeu Do you remember by any chance what was the issue exactly? We are facing the same problem, and cannot find a solution.

**Comment by icc-romeu** [comment]
Hi @mtapiero, as I explained before, it was a B2C configuration issue but I am not the Azure engineer and I do not have the details.

We have been using the react-native-msal library and I found the same problem and this time, the thing was that the idToken is returned in a different place if the account does not belong to the local tenant. Maybe it helps: https://github.com/stashenergy/react-native-msal/issues/109#issuecomment-915994382

**Comment by rcjoiner** [comment]
Just in case anyone else ends up here this might be the config problem you are looking for, there are other causes as well but we lost a lot of time finding this one out. 

We had this same issue and it was down to the fact we were returning the tenantId claim. Once this was removed from the RelyingParty it worked. We also removed the following line from the base template

`<OutputClaim ClaimTypeReferenceId="tenantId" PartnerClaimType="tid" />`

MS Support told us
"The existing MSAL caching code didn't anticipate the presence of tid claim in the token and therefore when this claim is present then it leads to the token being cached slightly differently by MSAL which then leads to cache miss on the subsequent silent token requests. "
